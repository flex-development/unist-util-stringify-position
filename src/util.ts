/**
 * @file util
 * @module unist-util-stringify-position/util
 */

import type { PointLike, PositionLike, Range } from './types'
import type State from './types/state'

/**
 * Standardize a column, line, or offset.
 *
 * > ðŸ‘‰ If `offset` is `true`, `value` is expected to already be a number.
 *
 * @internal
 *
 * @param {number | null | undefined} value - Column, line, or offset
 * @param {boolean | null | undefined} [offset] - `value` is offset?
 * @return {number} Column, line, or offset
 */
function index(
  value: number | null | undefined,
  offset: boolean | null | undefined
): number {
  return offset ? value! : value && typeof value === 'number' ? value : 1
}

/**
 * Serialize a point.
 *
 * @internal
 *
 * @param {PointLike | null | undefined} point - Point to serialize
 * @return {string} Pretty printed `point`
 */
function point(point: PointLike | null | undefined): string {
  return index(point?.line, false) + ':' + index(point?.column, false)
}

/**
 * Serialize a position.
 *
 * @internal
 *
 * @param {PositionLike | null | undefined} pos - Position to serialize
 * @param {State} state - Info passed around
 * @return {string} Pretty printed `position`
 */
function position(pos: PositionLike | null | undefined, state: State): string {
  /**
   * Serialized position.
   *
   * @var {string} serialized
   */
  let serialized: string = point(pos?.start) + '-' + point(pos?.end)

  if (
    state.offsets &&
    typeof pos?.end?.offset === 'number' &&
    typeof pos.start?.offset === 'number'
  ) {
    serialized += ', ' + index(pos.start.offset, state.offsets) + '-'
    serialized += index(pos.end.offset, state.offsets)
  }

  return serialized
}

/**
 * Serialize a range.
 *
 * @internal
 *
 * @param {Range | null | undefined} range - Range to serialize
 * @param {State} state - Info passed around
 * @return {string} Pretty printed `range`
 */
function range(range: Range | null | undefined, state: State): string {
  return position({ end: range?.[1], start: range?.[0] }, state)
}

export { point, position, range }
