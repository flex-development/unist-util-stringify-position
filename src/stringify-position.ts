/**
 * @file stringifyPosition
 * @module unist-util-stringify-position/stringifyPosition
 */

import type {
  Info,
  NodeLike,
  Options,
  PointLike,
  PositionLike,
  Range
} from './types'
import type State from './types/state'
import { point, position, range } from './util'

/**
 * Serialize the positional info of a node, point, position, or range.
 *
 * The serialized info is returned in one the following formats:
 *
 *  - `ls:cs-le:ce, os-oe` (node, position, range)
 *  - `ls:cs-le:ce` (node, position, range)
 *  - `l:c` (point)
 *
 * where `l` stands for line, `c` for column, `o` for offset, `s` for `start`,
 * and `e` for end.
 *
 * An empty string (`''`) is returned if the given info is neither a node,
 * point, position, or range.
 *
 * @see {@linkcode Info}
 * @see {@linkcode Options}
 *
 * @param {(Info | null)?} [info] - Node, point, position, or range
 * @param {(Options | null)?} [options] - Configuration options
 * @param {(boolean | null)?} [options.offsets=false] - Serialize offsets if
 * `info` is a node, position, or range
 * @return {string} Pretty printed positional info
 */
function stringifyPosition(
  info?: Info | null,
  options?: Options | null
): string

/**
 * Serialize the positional info of a node, point, position, or range.
 *
 * The serialized info is returned in one the following formats:
 *
 *  - `ls:cs-le:ce, os-oe` (node, position, range)
 *  - `ls:cs-le:ce` (node, position, range)
 *  - `l:c` (point)
 *
 * where `l` stands for line, `c` for column, `o` for offset, `s` for `start`,
 * and `e` for end.
 *
 * An empty string (`''`) is returned if the given info is neither a node,
 * point, position, or range.
 *
 * @see {@linkcode Options}
 *
 * @param {unknown?} [info] - Node, point, position, or range
 * @param {(Options | null)?} [options] - Configuration options
 * @param {(boolean | null)?} [options.offsets=false] - Serialize offsets if
 * `info` is a node, position, or range
 * @return {string} Pretty printed positional info
 */
function stringifyPosition(
  info?: unknown,
  options?: Options | null
): string

/**
 * Serialize the positional info of a node, point, position, or range.
 *
 * The serialized info is returned in one the following formats:
 *
 *  - `ls:cs-le:ce, os-oe` (node, position, range)
 *  - `ls:cs-le:ce` (node, position, range)
 *  - `l:c` (point)
 *
 * where `l` stands for line, `c` for column, `o` for offset, `s` for `start`,
 * and `e` for end.
 *
 * An empty string (`''`) is returned if the given info is neither node, point,
 * position, nor range.
 *
 * @see {@linkcode Info}
 * @see {@linkcode Options}
 *
 * @param {unknown?} [info] - Node, point, position, or range
 * @param {(Options | null)?} [options] - Configuration options
 * @param {(boolean | null)?} [options.offsets=false] - Serialize offsets if
 * `info` is a node, position, or range
 * @return {string} Pretty printed positional info
 */
function stringifyPosition(
  info?: unknown,
  options?: Options | null
): string {
  /**
   * Serialized positional info.
   *
   * @var {string} serialized
   */
  let serialized: string = ''

  if (typeof info === 'object' && info) {
    /**
     * Info passed around.
     *
     * @const {State} state
     */
    const state: State = { offsets: options?.offsets ?? false }

    // range
    if (Array.isArray(info)) serialized = range(<Range>info, state)

    // node
    if ('position' in info || 'type' in info) {
      serialized = position((<NodeLike>info).position, state)
    }

    // position
    if ('end' in info || 'start' in info) {
      serialized = position(<PositionLike>info, state)
    }

    // point
    if ('column' in info || 'line' in info) serialized = point(<PointLike>info)
  }

  return serialized
}

export default stringifyPosition
