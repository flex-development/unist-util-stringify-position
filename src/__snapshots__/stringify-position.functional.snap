// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`functional:stringifyPosition > node > should serialize node (invalid position #1) 1`] = `"1:1-1:1"`;

exports[`functional:stringifyPosition > node > should serialize node (invalid position #2) 1`] = `"1:1-1:1"`;

exports[`functional:stringifyPosition > node > should serialize node (invalid position #3) 1`] = `"1:1-1:1"`;

exports[`functional:stringifyPosition > node > should serialize node (no position) 1`] = `"1:1-1:1"`;

exports[`functional:stringifyPosition > node > should serialize node (valid) 1`] = `"63:8-63:55"`;

exports[`functional:stringifyPosition > node > should serialize node (valid, offsets) 1`] = `"114:76-115:1, 2606-2607"`;

exports[`functional:stringifyPosition > point > should serialize point (column only) 1`] = `"1:3"`;

exports[`functional:stringifyPosition > point > should serialize point (line only) 1`] = `"13:1"`;

exports[`functional:stringifyPosition > point > should serialize point (no indices) 1`] = `"1:1"`;

exports[`functional:stringifyPosition > point > should serialize point (valid) 1`] = `"2:3"`;

exports[`functional:stringifyPosition > position > should serialize position (invalid points #1) 1`] = `"1:1-1:1"`;

exports[`functional:stringifyPosition > position > should serialize position (invalid points #2) 1`] = `"1:1-1:1"`;

exports[`functional:stringifyPosition > position > should serialize position (invalid points #3) 1`] = `"1:1-1:1"`;

exports[`functional:stringifyPosition > position > should serialize position (no points) 1`] = `"1:1-1:1"`;

exports[`functional:stringifyPosition > position > should serialize position (valid) 1`] = `"111:6-128:42"`;

exports[`functional:stringifyPosition > position > should serialize position (valid, offsets) 1`] = `"1:1-4:4, 0-73"`;

exports[`functional:stringifyPosition > range > should serialize range (no points) 1`] = `"1:1-1:1"`;

exports[`functional:stringifyPosition > range > should serialize range (valid) 1`] = `"42:6-38:3"`;

exports[`functional:stringifyPosition > range > should serialize range (valid, offsets) 1`] = `"13:4-13:51, 347-521"`;
